core/executor/src/wasm_runtimes_cache.rs

```diff
@@ -0,0 +1,316 @@
+	instance: WasmModuleInstanceRef,
+	/// Runtime version according to `Core_version`.
+	///
+	/// Can be `None` if the runtime doesn't expose this function.
+	version: Option<RuntimeVersion>,
+	/// The snapshot of the instance's state taken just after the instantiation.
+	state_snapshot: StateSnapshot,
+}
+
+/// A state snapshot of an instance taken just after instantiation.
+///
+/// It is used for restoring the state of the module after execution.
+#[derive(Clone)]
+struct StateSnapshot {
+	/// The offset and the contents of the memory segments that should be copied at
+	/// to restore the snapshot.
+	data_segments: Vec<(u32, Vec<u8>)>,
+	/// The list of all global variables of the module in their sequential order.
+	global_mut_values: Vec<RuntimeValue>,
+	heap_pages: u32,
+}
+
+impl StateSnapshot {
+	// Returns `None` if instance is not valid.
+	fn take(
+		module_instance: &WasmModuleInstanceRef,
+		data_segments: Vec<DataSegment>,
+		heap_pages: u32,
+	) -> Option<Self> {
+		let mut prepared_segments = Vec::with_capacity(data_segments.len());
+		for mut segment in data_segments {
+			// Just replace contents of the segment since the segments will be discarded later
+			// anyway.
+			let contents = mem::replace(segment.value_mut(), vec![]);
+
+			let init_expr = segment.offset().code();
+			// [op, End]
+			if init_expr.len() != 2 {
+				return None;
+			}
+			let offset = match init_expr[0] {
+				Instruction::I32Const(v) => v as u32,
+				Instruction::GetGlobal(idx) => {
+					let global_val = module_instance.globals().get(idx as usize)?.get();
+					match global_val {
+						RuntimeValue::I32(v) => v as u32,
+						_ => return None,
+					}
+				}
+				_ => return None,
+			};
+			prepared_segments.push((offset, contents))
+		}
+
+		// Collect all values of mutable globals.
+		let global_mut_values = module_instance
+			.globals()
+			.iter()
+			.filter(|g| g.is_mutable())
+			.map(|g| g.get())
+			.collect();
+
+		Some(Self {
+			data_segments: prepared_segments,
+			global_mut_values,
+			heap_pages,
+		})
+	}
+
+	/// Reset the runtime instance to the initial version by restoring
+	/// the preserved memory and globals.
+	///
+	/// Returns `Err` if applying the snapshot is failed.
+	fn apply(&self, instance: &WasmModuleInstanceRef) -> Result<(), CacheError> {
+		let memory = instance
+			.export_by_name("memory")
+			.ok_or(CacheError::ApplySnapshotFailed)?
+			.as_memory()
+			.cloned()
+			.ok_or(CacheError::ApplySnapshotFailed)?;
+
+		// First, erase the memory and copy the data segments into it.
+		memory
+			.erase()
+			.map_err(|_| CacheError::ApplySnapshotFailed)?;
+		for (offset, contents) in &self.data_segments {
+			memory
+				.set(*offset, contents)
+				.map_err(|_| CacheError::ApplySnapshotFailed)?;
+		}
+
+		// Second, restore the values of mutable globals.
+		for (global_ref, global_val) in instance
+			.globals()
+			.iter()
+			.filter(|g| g.is_mutable())
+			.zip(self.global_mut_values.iter())
```

Comment:
We could here just use a `try_for_each` and drop the last `Ok(())`.

Reply:
Ah, nice to know about `try_for_each`. However, I personally prefer good ol' for loops instead of combinator chains. The reason is one should be really careful using them when the chain of combinators becomes long and nested.

IMO combinators fail to convey the deep nesting compared to plain old control flow structures. In this particular case, they will also introduce additional indent which can harm the comments readability. 

That said, I don't feel strongly about this, if you do I can change it!